src/python/pants/engine/internals/mapper.py

```diff
@@ -152,44 +151,36 @@ def __repr__(self) -> str:
         )
 
 
-@frozen_after_init
-@dataclass(unsafe_hash=True)
+@dataclass(frozen=True)
 class SpecsFilter:
     """Filters addresses with the `--tags` and `--exclude-target-regexp` options."""
 
-    tags: tuple[str, ...]
-    exclude_target_regexps: tuple[str, ...]
-
-    def __init__(
-        self,
-        *,
-        tags: Iterable[str] | None = None,
-        exclude_target_regexps: Iterable[str] | None = None,
-    ) -> None:
-        self.tags = tuple(tags or [])
-        self.exclude_target_regexps = tuple(exclude_target_regexps or [])
+    is_specified: bool
+    tags_filter: TargetFilter
+    exclude_target_regexps_filter: TargetFilter
 
-    @memoized_property
-    def _exclude_regexps(self) -> tuple[Pattern, ...]:
-        return tuple(re.compile(pattern) for pattern in self.exclude_target_regexps)
+    @classmethod
+    def create(cls, *, tags: Iterable[str], exclude_target_regexps: Iterable[str]) -> SpecsFilter:
```

Comment:
You could also override `__new__` for this.

Reply:
Indeed, but that is a bit magical. The `Specs` refactor has prompted me to be more in favor of a `@dataclass` storing as plain-and-simple fields its normalized data, and using `.create()` to do special setup.

`@frozen_after_init` is an antipattern I think, I kind of regret introducing it.

(My thoughts are somewhat inspired by Rust design patterns)

Reply:
Overall I think I agree.

And unlike `Specs`, it is almost always going to be a bug to directly create a `SpecsFilter` without having called `.create()`. 

I'll try this out. Thanks!

Reply:
Hm I don't think overridding `__new__` is very well blessed when you already have a frozen dataclass. Dataclass is doing some fancy things with `__new__` and `__init__` from what I understand.

I came up with this diff

```diff
diff --git a/src/python/pants/engine/internals/mapper.py b/src/python/pants/engine/internals/mapper.py
index 652b7f790..46fc4b447 100644
--- a/src/python/pants/engine/internals/mapper.py
+++ b/src/python/pants/engine/internals/mapper.py
@@ -159,8 +159,7 @@ class SpecsFilter:
     tags_filter: TargetFilter
     exclude_target_regexps_filter: TargetFilter
 
-    @classmethod
-    def create(cls, *, tags: Iterable[str], exclude_target_regexps: Iterable[str]) -> SpecsFilter:
+    def __new__(cls, *, tags: Iterable[str], exclude_target_regexps: Iterable[str]) -> SpecsFilter:
         exclude_patterns = tuple(re.compile(pattern) for pattern in exclude_target_regexps)
 
         def exclude_target_regexps_filter(tgt: Target) -> bool:
@@ -174,11 +173,11 @@ class SpecsFilter:
 
         tags_filter = and_filters(create_filters(tags, tags_outer_filter))
 
-        return SpecsFilter(
-            is_specified=bool(tags or exclude_target_regexps),
-            tags_filter=tags_filter,
-            exclude_target_regexps_filter=exclude_target_regexps_filter,
-        )
+        obj = super().__new__()
+        obj.is_specified = bool(tags or exclude_target_regexps)
+        obj.tags_filter = tags_filter
+        obj.exclude_target_regexps_filter = exclude_target_regexps_filter
+        return obj
 
     def matches(self, target: Target) -> bool:
         """Check that the target matches the provided `--tag` and `--exclude-target-regexp`
diff --git a/src/python/pants/engine/internals/mapper_test.py b/src/python/pants/engine/internals/mapper_test.py
index 42d622d32..30e0b6359 100644
--- a/src/python/pants/engine/internals/mapper_test.py
+++ b/src/python/pants/engine/internals/mapper_test.py
@@ -131,7 +131,7 @@ def test_address_family_duplicate_names() -> None:
 
 
 def test_specs_filter() -> None:
-    specs_filter = SpecsFilter.create(tags=["-a", "+b"], exclude_target_regexps=["skip-me"])
+    specs_filter = SpecsFilter(tags=["-a", "+b"], exclude_target_regexps=["skip-me"])
 
     class MockTgt(Target):
         alias = "tgt"
diff --git a/src/python/pants/engine/internals/specs_rules.py b/src/python/pants/engine/internals/specs_rules.py
index 53ef2601d..48b7d880e 100644
--- a/src/python/pants/engine/internals/specs_rules.py
+++ b/src/python/pants/engine/internals/specs_rules.py
@@ -232,7 +232,7 @@ def filter_targets(targets: Targets, specs_filter: SpecsFilter) -> FilteredTarge
 
 @rule
 def setup_specs_filter(global_options: GlobalOptions) -> SpecsFilter:
-    return SpecsFilter.create(
+    return SpecsFilter(
         tags=global_options.tag, exclude_target_regexps=global_options.exclude_target_regexp
     )

```

```
    return SpecsFilter(
  File "/Users/ericarellano/code/pants/src/python/pants/engine/internals/mapper.py", line 176, in __new__
    obj = super().__new__()
TypeError: object.__new__(): not enough arguments
```

Reply:
Mm. That is because `__new__` takes the `cls` as an argument. But I tried it, and you're right that `dataclass` makes this challenging. The last time I had tried it was with `namedtuple`.